File: Java-CodeServer/src/main/resources/js/prompto/value/Range.js
var Value = require("./Value").Value;
var Integer = require("./Integer").Integer;
var IndexOutOfRangeError = require("../error/IndexOutOfRangeError").IndexOutOfRangeError;
var BaseType = require("../type/BaseType").BaseType;
var RangeType = require("../type/RangeType").RangeType;

function Range(itemType, left, right) {
    if(!(itemType instanceof BaseType))
        return;
	Value.call(this, new RangeType(itemType));
	var cmp = left.cmp(right);
	if(cmp<0) {
		this.low = left;
		this.high = right;
	} else {
		this.low = right;
		this.high = left;
	}
	return this;
}

Range.prototype = Object.create(Value.prototype);
Range.prototype.constructor = Range;

Range.prototype.toString = function() {
	return "[" + (this.low==null?"":this.low.toString()) + ".."
			+ (this.high==null?"":this.high.toString()) + "]";
};

Range.prototype.equals = function(obj) {
	if(obj instanceof Range) {
		return this.low.equals(obj.low) && this.high.equals(obj.high);
	} else {
		return false;
	}
};


Range.prototype.hasItem = function(context, lval) {
	var a = lval.cmp(this.low);
	var b = this.high.cmp(lval);
	return a>=0 && b>=0;
};

Range.prototype.getItemInContext = function(context, index) {
	if (index instanceof Integer) {
		try {
			var value = this.getItem(index.IntegerValue());
			if (value instanceof Value) {
				return value;
			} else {
				throw new InternalError("Item not a value!");
			}
		} catch (e) {
			throw new IndexOutOfRangeError();
		}

	} else {
		throw new SyntaxError("No such item:" + index.toString());
	}
};

Range.prototype.slice = function(fi, li) {
	var size = this.size();
	var first = this.checkFirst(fi, size);
	var last = this.checkLast(li, size);
	return this.newInstance(this.getItem(first),this.getItem(last));
}

Range.prototype.checkFirst = function(fi, size) {
	var value = (fi == null) ? 1 : fi.IntegerValue();
	if (value < 1 || value > size) {
		throw new IndexOutOfRangeError();
	}
	return value;
};

Range.prototype.checkLast = function(li, size) {
	var value = (li == null) ? size : li.IntegerValue();
	if (value < 0) {
		value = size + 1 + li.IntegerValue();
	}
	if (value < 1 || value > size) {
		throw new IndexOutOfRangeError();
	}
	return value;
};

/*
 @Override
 public Iterable<T> getItems(Context context) {
 return new RangeIterable(context);
 }

*/


Range.prototype.getIterator = function(context) {
	return new RangeIterator(context, this);
};

function RangeIterator(context, range) {
	this.context = context;
	this.range = range;
	this.index = 0;
	return this;
}

RangeIterator.prototype.hasNext = function() {
	return this.index<this.range.size();
};

RangeIterator.prototype.next = function() {
	return this.range.getItemInContext(this.context, new Integer(++this.index));
};

/*
@Override
public boolean isEmpty() {
	return size()==0;
}

public abstract long size();
public abstract int compare(T o1,T o2);
public abstract T getItem(long index);
public abstract Range<T> newInstance(T left,T right);

*/

exports.Range = Range;

File: Java-Core/src/main/java/prompto/value/Range.java
package prompto.value;

import java.util.Iterator;

import prompto.compiler.CompilerUtils;
import prompto.compiler.Flags;
import prompto.compiler.IOperand;
import prompto.compiler.MethodConstant;
import prompto.compiler.MethodInfo;
import prompto.compiler.Opcode;
import prompto.compiler.ResultInfo;
import prompto.error.IndexOutOfRangeError;
import prompto.error.PromptoError;
import prompto.error.SyntaxError;
import prompto.expression.IExpression;
import prompto.intrinsic.PromptoRange;
import prompto.runtime.Context;
import prompto.type.IType;
import prompto.type.RangeType;

public abstract class Range<T extends IValue> extends BaseValue implements IContainer<T>, IRange<T> {
	
	PromptoRange<T> range;
	
	public Range(IType type, T left, T right) {
		super(new RangeType(type));
		// can't just use T extends Comparable<T> because T may already extend Comparable<R> with R!=T
		int cmp = compare(left,right);
		if(cmp<0)
			range = new PromptoRange<T>(left, right);
		else
			range = new PromptoRange<T>(right, left);
	}
	
	@Override
	public String toString() {
		return range.toString();
	}
	
	public T getLow() {
		return range.getLow();
	}
	
	public T getHigh() {
		return range.getHigh();
	}

	@Override
	public boolean equals(Object obj) {
		return obj instanceof Range && range.equals(((Range<?>)obj).range);
	}
		
	public static ResultInfo compileEquals(Context context, MethodInfo method, ResultInfo left, IExpression exp, Flags flags) throws SyntaxError {
		exp.compile(context, method, flags);
		IOperand oper = new MethodConstant(
				PromptoRange.class, 
				"equals",
				Object.class, boolean.class);
		method.addInstruction(Opcode.INVOKEVIRTUAL, oper);
		if(flags.isReverse()) {
			// perform 1-0
			method.addInstruction(Opcode.ICONST_1);
			method.addInstruction(Opcode.SWAP);
			method.addInstruction(Opcode.ISUB);
		}
		if(flags.toNative())
			return new ResultInfo(boolean.class, false);
		else
			return CompilerUtils.booleanToBoolean(method);
	}

	public boolean hasItem(Context context, IValue lval) {
		@SuppressWarnings("unchecked")
		T val = (T)lval;
		return compare(val, getLow())>=0 && compare(getHigh(), val)>=0;
	}
	
	@SuppressWarnings("unchecked")
	public T getItem(Context context, IValue index) throws PromptoError {
		if (index instanceof Integer) {
			try {
				Object value = this.getItem(((Integer) index).longValue());
				if (value instanceof IValue)
					return (T) value;
				else
					throw new InternalError("Item not a value!");
			} catch (IndexOutOfBoundsException e) {
				throw new IndexOutOfRangeError();
			}

		} else
			throw new SyntaxError("No such item:" + index.toString());
		  			
	}

	public Range<T> slice(Integer fi, Integer li) throws PromptoError {
		long size = length();
		long _fi = fi==null ? 1L : fi.longValue();
		if(_fi<0)
			throw new IndexOutOfRangeError();
		long _li = li==null ? size : li.longValue();
		if(_li<0)
			_li = size + 1 + _li;
		else if(_li>size)
			throw new IndexOutOfRangeError();
		return newInstance(getItem(_fi),getItem(_li));
	}

	@Override
	public Iterable<T> getIterable(Context context) {
		return new RangeIterable(context);
	}	
	
	class RangeIterable implements Iterable<T> {
		
		Context context;
		
		public RangeIterable(Context context) {
			this.context = context;
		}

		@Override
		public Iterator<T> iterator() {
			return new RangeIterator();
		}

		class RangeIterator implements Iterator<T> {
		
			long index = 0L;
			
			@Override
			public boolean hasNext() {
				return index<length();
			}
	
			@Override
			public T next() {
				try {
					return getItem(context, new Integer(++index));
				} catch(Throwable t) {
					throw new InternalError(t.getMessage());
				}
			}
	
			@Override
			public void remove() {
				throw new RuntimeException("Shold never get there!");
			}
			
		}
	}
	
	@Override
	public boolean isEmpty() {
		return length()==0;
	}
	
	public abstract long length();
	public abstract int compare(T o1,T o2);
	public abstract T getItem(long index);
	public abstract Range<T> newInstance(T left,T right);



}
